### ISSUE TITLE
Improve StringName and NodePath usability

### ISSUE THREAD

**Issue by KoBeWi** [godotengine/godot-proposals#4721]
### Describe the project you are working on

A very big project I am converting to 4.0

### Describe the problem or limitation you are having in your project

Godot 4.0 added StringNames as a type usable from GDScript. Many methods now expect StringNames (implicit String conversion is available though), but StringNames are also nice for custom methods that rely a lot on string comparisons, but don't modify the strings. So my code keeps getting new StringNames everywhere.

Another thing are NodePaths, which are used in some methods too. I neglected them so far, because when you use String it will just get automatically converted when needed, but using correct type makes the code faster and it doesn't hurt to use it.

My only problem is that syntax-wise StringNames and NodePaths are basically Strings with extra symbol at the beginning. This is nice for converting project, because you can just slap `&` to your strings, but not very convenient for new code and also it's less readable (mostly due to being colored the same as Strings and the extra symbol standing out).

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

We have a special coloring for `$` operator:
![image](https://user-images.githubusercontent.com/2223172/175011126-016c1ddb-737d-48b9-98dd-5df14edcb486.png)

It should be the same for `&` and `^`, i.e. StringName and NodePath could get their own unique color, so instead of
![image](https://user-images.githubusercontent.com/2223172/175011343-e3857779-cec6-4719-8755-4c57310289cf.png)
we could have
![image](https://user-images.githubusercontent.com/2223172/175012420-0204af12-2301-4c5b-be14-731c035ff067.png)
(ignore that the symbol is darker; it should be the same)

And since `%` got it's special syntax (`%UniqueNode`), we could allow unquoted StringNames and NodePaths too:
![image](https://user-images.githubusercontent.com/2223172/175012512-9a6d8247-1902-4d42-80a0-710f5895f399.png)

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

Allow this and make it uniquely-colored in the script editor:
```GDScript
&StringName
^NodePath
```

### If this enhancement will not be used often, can it be worked around with a few lines of script?

Syntax part can be worked around by using quotes 🙄 but the custom highlighting part not.

### Is there a reason why this should be core and not an add-on in the asset library?

StringName and NodePath are core types.

**Comment by YuriSizov** [comment]
> (ignore that the symbol is darker; it should be the same)

But I kinda like it... 👀

**Comment by KoBeWi** [comment]
tbh I don't like the idea of darker symbols. It's just unnecessary. Although with quotes maybe makes sense; you get a different-colored string and actually the symbol becomes redundant. Or something.

> It starts to remind me of Ruby, and not in a good way.

Ruby is great tho. I don't see any problem in skipping quotes; they are useless.

> this may add a lot more colors for third-party syntax themes to support

Why is this even relevant xd They can just set the same color for every option.

**Comment by SilencedPerson** [comment]
> > (ignore that the symbol is darker; it should be the same)
> 
> But I kinda like it... 👀

I have to agree on this, I would love this on the $ syntax, I tend to avoid it, because the symbol makes it very hard to read, the darker version remedies this substantially.

**Comment by MewPurPur** [comment]
This can be closed, right?

**Comment by KoBeWi** [comment]
No, the second part of the proposal (unquoted StringNames and NodePaths) is still not addressed.

**Comment by MewPurPur** [comment]
Ah mhm. Not too fond of the idea personally, I think it's nice to be clear those are just special types of strings and not allow alternate syntax as it may be confusing. Not just that, but the quotation marks show where the variant starts and ends - without it, special symbols would screw with NodePath. `/` is particularly problematic (does it represent another node, or a division by a var?) and not having this oddity would require an invasive exception.

StringName plays nicely with most of the ways it's used in the editor (a few are still problematic), but I personally wouldn't use it without quotation marks. For example it's a common practice to use special symbols for translatable strings, which get passed to `tr(StringName)`

**Comment by Mickeon** [comment]
The "second part" of the proposal was mentioned years later in 
- https://github.com/godotengine/godot-proposals/discussions/10854

and Dalexeev's sentiment is pretty much identical.
